clippy_utils/src/ty.rs

```diff
@@ -590,3 +591,133 @@ pub fn is_c_void(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
         false
     }
 }
+
+/// Makes the projection type for the named associated type in the given impl or trait impl.
+///
+/// This function is for associated types which are "known" to exist, and as such, will only return
+/// `None` when debug assertions are disabled in order to prevent ICE's. With debug assertions
+/// enabled this will check that the named associated type exists, the correct number of
+/// substitutions are given, and that the correct kinds of substitutions are given (lifetime,
+/// constant or type). This will not check if type normalization would succeed.
+pub fn make_projection<'tcx>(
```

Comment:
I must admit that I don't quite understand what this is about.

Reply:
If some type implements a `T`, then the projected type `T::Ty` will exist. This is the intended use of the function. It's possible while writing lints to mix things up and make an invalid assumption about this, or to mess up the generic substitutions. In that case `evaluate_obligations` will fail or crash when it gets called.

Both `make_projection` and `make_normalized_projection` will crash on debug builds so the error is obvious on debug builds. For released version of clippy this will avoid crashing and instead risk resulting in false positives or negatives. Running to the end with slightly wrong results is generally more useful than crashing part way through.